README.md
# yet_another_carousel
 
![gif](misc/20191020204855.gif)


## donate

[Buy me a ☕](https://www.paypal.com/paypalme2/JGarza9788/3.50)



## Parameter
```dart
  //this is the list of widgets that will be rendered
  List<Widget> children;

  //this is the curve that will be used for positioning
  Curve positionCurve;

  //this is the curve that will be used for scaling the widgets
  Curve scaleCurve;

  //this is the curve that will be used while transitioning out of screen (left side)
  Curve outCurve;

  //this is the color of the background
  Color backgroundColor;

  //if the widgets will fade out
  bool fadeOut;
  
  //if the widgets will scale out
  bool scaleOut;
  
  //if the carousel will roll to the nearest widget after drag
  bool rollToNearest;
  
  //the width of the carousel
  double width;
  
  //the height of the carousel
  double height;
  
  //the width of the widgets (anything under 1 will be treated as a ratio)
  double widgetWidth;

  //the height of the widgets (anything under 1 will be treated as a ratio)
  double widgetHeight;
  
  //this will scale the widget horizontally
  double widthFactor;
  
  //this can shift the widget on the x axis
  double xPosition;
  
  //this is the offset of the carousel on the right hand side
  Offset tailOfLineOffset;
  
  //this is the offset of the carousel on the left hand side
  Offset headOfLineOffset;
  
  //the widget will scroll to this index after change.
  int scrollTo;
```


## Example

```dart
Carousel(
    children: theseIcons,
    positionCurve: Curves.easeOut,
    scaleCurve: Curves.easeOut,
    fadeOut: true,
    scaleOut: true,
    widthFactor: 2.0,
    scrollTo: wIndex,
),
```

CHANGELOG.md
## 0.1.2

* Minor updates

## 0.1.1

* Minor updates
  
## 0.1.0

* Initial Open Source release.

lib/Carousel.dart
import 'dart:math';

import 'package:flutter/material.dart';


class Carousel extends StatefulWidget {
  Carousel({
    @required this.children,
    this.positionCurve = Curves.linear,
    this.scaleCurve = Curves.linear,
    this.outCurve = Curves.linear,
    this.backgroundColor = Colors.grey,
    this.fadeOut = false,
    this.scaleOut = false,
    this.rollToNearest =  true,
    this.width = double.infinity,
    this.height = 250.0,
    this.widgetWidth = 150.0,
    this.widgetHeight = 150.0,
    this.widthFactor = 1.0,
    this.xPosition = 0.0,
    this.tailOfLineOffset = const Offset(0.0, 0.0),
    this.headOfLineOffset = const Offset(-1.0, -1.0),
    this.scrollTo = 0,
  });

  List<Widget> children;
  Curve positionCurve;
  Curve scaleCurve;
  Curve outCurve;
  Color backgroundColor;
  Size widgetSize;
  bool fadeOut;
  bool scaleOut;
  bool rollToNearest;
  double width;
  double height;
  double widgetWidth;
  double widgetHeight;
  double widthFactor;
  double xPosition;
  Offset tailOfLineOffset;
  Offset headOfLineOffset;
  int scrollTo;


  @override
  _CarouselState createState() => _CarouselState();
}

class _CarouselState extends State<Carousel> with SingleTickerProviderStateMixin{

  GlobalKey _key = GlobalKey();

  double _shift = 0.0;
  int _scrollTo = 0;

  double _velocity = 0.0;
  int _dragEndTime = 0;

  AnimationController _controller;

  List<Widget> wList = [];

  states state = states.idle;


  double lastWidgetPosition = 0.0;
  List<Widget> refresh()
  {
    List<Widget> finalWidgetList = [];

    if (_size == Size.zero)
    {
      return finalWidgetList;
    }

    var params = [];

    for(int i = 0; i < widget.children.length;i++)
    {
      double r = i/widget.children.length ;
//      double rawPos = ratio * _sizeWithExtra.width + _shift;
      double ratio = (r + _shift)%1.0;
      _scrollValue = ratio;
//      double pos = ratio * _sizeWithExtra.width;

      params.add(
          {
            "index":i,
            "ratio":ratio,
          }
          );
    }

    params.sort( (a,b) =>  a["ratio"] < b["ratio"] ? 1:0  );
//    print(params);
//    print('\n');
//    print(params[0]);

    if (state != states.rollToNearest)
      {
        _activeWidgetIndex = params[params.length-1]['index'] ;

//        print(params[params.length-1]);
//        print(params[params.length-2]);
//
//        if (params[params.length-1]['ratio'] > 0.009)
//          {
//            _activeWidgetIndex = params[params.length-2]['index'] ;
//          }
      }

//    print(state);
//    print(_activeWidgetIndex);

    for (int i = 0; i < params.length;i++)
    {
//      print(params[i]);


      double r =  params[i]["ratio"];

      double rightEnd = widget.outCurve.transform(r) * 25.0;
      rightEnd = rightEnd.clamp(0.0, 1.0);

      double leftEnd = 1.0 - (widget.outCurve.transform(r) * 25);
      leftEnd = leftEnd.clamp(0.0, 1.0);

//      print(leftEnd);

      double pos = widget.positionCurve.transform(r);
      pos = ((pos * widget.widthFactor) - widget.xPosition) - (1-rightEnd) ;
      pos += widget.tailOfLineOffset.dx  - ( (1-r) * widget.tailOfLineOffset.dx);

      pos += ( (widget.headOfLineOffset.dx + 1.0) * leftEnd);
      double scale  = 1 - widget.scaleCurve.transform(r);
      scale = widget.scaleOut ? scale * (leftEnd + 1.0) : scale;

      double maxWidth = widget.widgetWidth <= 1.0 ? (_size.width * widget.widgetWidth) : widget.widgetWidth;
      double maxHeight = widget.widgetHeight <= 1.0 ? (_size.height * widget.widgetHeight) : widget.widgetHeight;


      double posY = widget.tailOfLineOffset.dy  - ( (1-r) * widget.tailOfLineOffset.dy);
      posY += ( (widget.headOfLineOffset.dy + 1.0) * (leftEnd + 1.0));
      double opacity = widget.fadeOut ? rightEnd: 1.0;


      finalWidgetList.add(
        Align(
          alignment: Alignment(
            pos,
            posY
          ),
          child: Transform.scale(
            scale: scale,
            child: Opacity(
              opacity: opacity,
              child: FittedBox(
                fit: BoxFit.fitWidth,
                alignment: Alignment.center,
                child: Container(
                  height: maxHeight,
                  width: maxWidth,
                  child: widget.children[params[i]["index"]],
                ),
//                child: widget.children[params[i]["index"]],
              ),
            ),
          ),
        ),
      );

    }

//    finalWidgetList.add(
//      Align(
//        alignment: Alignment(
//          0.0,0.90
//        ),
//        child: Container(
//            height: 10.0,
//            width: _size.width * 0.80,
//            decoration: BoxDecoration(
//              color: Colors.blueGrey[900],
//              borderRadius: BorderRadius.circular(5.0),
//              boxShadow: [
//                new BoxShadow(
//                  color: Colors.black.withOpacity(0.50),
//                  offset: new Offset(2.5, 2.5),
//                  blurRadius: 10.0,
//                )
//              ],
//            ),
//            child: Stack(
//              overflow: Overflow.visible,
//              children: <Widget>[
//                Align(
//                  alignment: Alignment
//                    (
//                      (_scrollValue * 2.0) - 1.0
//                      ,0.0
//                  ),
//                  child: Container(
//                    height: 10.0,
//                    width: 50.0,
//                    decoration: BoxDecoration(
//                      color: Colors.green[900],
//                      borderRadius: BorderRadius.circular(5.0),
//                      boxShadow: [
//                        new BoxShadow(
//                          color: Colors.black.withOpacity(0.50),
//                          offset: new Offset(2.5, 2.5),
//                          blurRadius: 10.0,
//                        )
//                      ],
//                    ),
////                    child:
//                  ),
//                ),
//                Align(
//                  alignment: Alignment(
//                      (_scrollValue * 2.0) - 1.0
//                      ,0
//                  ),
//                  child: Text(
//                    '${(_scrollValue).toStringAsFixed(5)}',
//                    style: TextStyle(color: Colors.lightGreenAccent,fontSize: 10.0),
//                    softWrap: true,
//                  ),
//                ),
//              ],
//            ),
//          ),
//      )
//    );


    return finalWidgetList;
  }

  Size _size = Size.zero;
  _getSize(){
    RenderBox RB = _key.currentContext.findRenderObject();
    _size = RB.size;
  }

    @override
  void initState() {
    // TODO: implement initState
    WidgetsBinding.instance.addPostFrameCallback(_afterLayout);
    super.initState();


    refresh();

    _controller = AnimationController(
      duration: Duration(milliseconds: 5000),
      vsync: this,
    );

    _controller.addStatusListener((status){
      print(status);
      //moves between forwards and back

      if(status == AnimationStatus.completed)
      {
        print(state);
        if (state == states.drag)
          {
            state = states.dragEnd;
            _startRollToNearest();
          }
        else if (state == states.dragEnd)
          {
            _startRollToNearest();
          }
        else if (state != states.idle)
          {
            state = states.idle;
          }
      }

    });

    _controller.addListener((){
      print(state);
      setState(() {
//        print(_controller.value);

        if (state == states.idle)
          {
            //do nothing
          }
        else if (state == states.dragEnd)
          {
            _afterDrag();
          }
        else if (state == states.rollToNearest)
          {
            _rollToNearest();
          }
        else if (state == states.scrollTo)
          {
            _autoScroll();
          }


      });
    });
  }
  _afterLayout(_) {
    _getSize();
    refresh();
  }

  _afterDrag(){
//    print(_controller.value);
    int t = DateTime.now().millisecondsSinceEpoch - _dragEndTime;
    double addVel = _velocity * pow(0.99,t);
    _shift += addVel;

    print(addVel);
    if (addVel.abs() < 0.00001)
      {
        _startRollToNearest();
      }
  }

//  bool _isAutoScrolling = false;
  _autoScroll(){
//    print('_isAutoScrolling $_isAutoScrolling');
//    if (_isAutoScrolling)
//     {
       double nShift  = (_scrollTo/-widget.children.length) + 0.04;
       _shift = _shift + (nShift - _shift) * _controller.value;
//     }
//    print('_autoScrollTo $_scrollTo');
  }

  _startRollToNearest(){
    if (widget.rollToNearest)
    {
      state = states.rollToNearest;

      _shift = (((_shift + 1.0) % 2.0) -1.0);
//      _shift = 5.0;
//      _shift = _shift % 1.0;
//      print('_shift $_shift');

      _controller.stop(canceled: true);
      _controller.duration = Duration(milliseconds: 1000);
      _controller.forward(from: 0.0);

    }
  }


  _rollToNearest() {

    double nShift  = (_activeWidgetIndex/-widget.children.length) + 0.04;
    double dShift = (nShift-_shift);

    if (dShift.abs() > 0.95)
    {
      dShift += dShift.sign * -1.0;
    }

//    _shift = _shift + (nShift-_shift) *  _controller.value  ;
    double t = Curves.easeInOut.transform(_controller.value);
    _shift = _shift + dShift *  t  ;

  }


  @override
  Widget build(BuildContext context) {

    refresh();

    if (widget.scrollTo != _scrollTo  )
    {
      state = states.scrollTo;

      _scrollTo = widget.scrollTo;
      _controller.duration = Duration(milliseconds: 1000);
      _controller.forward(from: 0.0);

    }

    return GestureDetector(
      onHorizontalDragUpdate: (value){

        state = states.drag;

        setState(() {
          _shift += value.primaryDelta/_size.width;
        });

      },
      onHorizontalDragEnd: (value) {

        state = states.dragEnd;

        _dragEndTime = DateTime.now().millisecondsSinceEpoch;
        _velocity = value.primaryVelocity/1000;

        _controller.duration = Duration(milliseconds: (value.primaryVelocity).abs().round());
        _controller.forward(from: 0.0);

      },
      child: Container(
        key: _key,
        height: widget.height,
        width: widget.width,
        color: widget.backgroundColor,
        child: Stack(
          overflow: Overflow.clip,
          alignment: AlignmentDirectional.centerStart,
          children: refresh(),
        ),
      ),
    );
  }
}

double _scrollValue = 0.0;
double getScrollValue(){
  return _scrollValue;
}

int _activeWidgetIndex = 0;
int getActiveWidgetIndex(){
  return _activeWidgetIndex;
}

enum states{
  idle,
  drag,
  dragEnd,
  scrollTo,
  rollToNearest,
}


lib/friction_simulation.dart
// Copyright 2016 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'dart:math' as math;

//import 'simulation.dart';
//import 'tolerance.dart';

import 'package:flutter/physics.dart';

/// A simulation that applies a drag to slow a particle down.
///
/// Models a particle affected by fluid drag, e.g. air resistance.
///
/// The simulation ends when the velocity of the particle drops to zero (within
/// the current velocity [tolerance]).
class FrictionSimulation extends Simulation {
  /// Creates a [FrictionSimulation] with the given arguments, namely: the fluid
  /// drag coefficient, a unitless value; the initial position, in the same
  /// length units as used for [x]; and the initial velocity, in the same
  /// velocity units as used for [dx].
  FrictionSimulation(
      double drag,
      double position,
      double velocity, {
        Tolerance tolerance = Tolerance.defaultTolerance,
      }) : _drag = drag,
        _dragLog = math.log(drag),
        _x = position,
        _v = velocity,
        super(tolerance: tolerance);

  /// Creates a new friction simulation with its fluid drag coefficient set so
  /// as to ensure that the simulation starts and ends at the specified
  /// positions and velocities.
  ///
  /// The positions must use the same units as expected from [x], and the
  /// velocities must use the same units as expected from [dx].
  ///
  /// The sign of the start and end velocities must be the same, the magnitude
  /// of the start velocity must be greater than the magnitude of the end
  /// velocity, and the velocities must be in the direction appropriate for the
  /// particle to start from the start position and reach the end position.
  factory FrictionSimulation.through(double startPosition, double endPosition, double startVelocity, double endVelocity) {
    assert(startVelocity == 0.0 || endVelocity == 0.0 || startVelocity.sign == endVelocity.sign);
    assert(startVelocity.abs() >= endVelocity.abs());
    assert((endPosition - startPosition).sign == startVelocity.sign);
    return FrictionSimulation(
      _dragFor(startPosition, endPosition, startVelocity, endVelocity),
      startPosition,
      startVelocity,
      tolerance: Tolerance(velocity: endVelocity.abs()),
    );
  }

  final double _drag;
  final double _dragLog;
  final double _x;
  final double _v;

  // Return the drag value for a FrictionSimulation whose x() and dx() values pass
  // through the specified start and end position/velocity values.
  //
  // Total time to reach endVelocity is just: (log(endVelocity) / log(startVelocity)) / log(_drag)
  // or (log(v1) - log(v0)) / log(D), given v = v0 * D^t per the dx() function below.
  // Solving for D given x(time) is trickier. Algebra courtesy of Wolfram Alpha:
  // x1 = x0 + (v0 * D^((log(v1) - log(v0)) / log(D))) / log(D) - v0 / log(D), find D
  static double _dragFor(double startPosition, double endPosition, double startVelocity, double endVelocity) {
    return math.pow(math.e, (startVelocity - endVelocity) / (startPosition - endPosition));
  }

  @override
  double x(double time) => _x + _v * math.pow(_drag, time) / _dragLog - _v / _dragLog;

  @override
  double dx(double time) => _v * math.pow(_drag, time);

  /// The value of [x] at `double.infinity`.
  double get finalX => _x - _v / _dragLog;

  /// The time at which the value of `x(time)` will equal [x].
  ///
  /// Returns `double.infinity` if the simulation will never reach [x].
  double timeAtX(double x) {
    if (x == _x)
      return 0.0;
    if (_v == 0.0 || (_v > 0 ? (x < _x || x > finalX) : (x > _x || x < finalX)))
      return double.infinity;
    return math.log(_dragLog * (x - _x) / _v + 1.0) / _dragLog;
  }

  @override
  bool isDone(double time) => dx(time).abs() < tolerance.velocity;
}

///// A [FrictionSimulation] that clamps the modeled particle to a specific range
///// of values.
//class BoundedFrictionSimulation extends FrictionSimulation {
//  /// Creates a [BoundedFrictionSimulation] with the given arguments, namely:
//  /// the fluid drag coefficient, a unitless value; the initial position, in the
//  /// same length units as used for [x]; the initial velocity, in the same
//  /// velocity units as used for [dx], the minimum value for the position, and
//  /// the maximum value for the position. The minimum and maximum values must be
//  /// in the same units as the initial position, and the initial position must
//  /// be within the given range.
//  BoundedFrictionSimulation(
//      double drag,
//      double position,
//      double velocity,
//      this._minX,
//      this._maxX,
//      ) : assert(position.clamp(_minX, _maxX) == position),
//        super(drag, position, velocity);
//
//  final double _minX;
//  final double _maxX;
//
//  @override
//  double x(double time) {
//    return super.x(time).clamp(_minX, _maxX);
//  }
//
//  @override
//  bool isDone(double time) {
//    return super.isDone(time) ||
//        (x(time) - _minX).abs() < tolerance.distance ||
//        (x(time) - _maxX).abs() < tolerance.distance;
//  }
//}

example/main.dart

import 'package:flutter/material.dart';
import 'package:yet_another_carousel/yet_another_carousel.dart';


//void main() => runApp(SwipeDemoApp());
void main() => runApp(MyApp());


class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {


  int wIndex = 0;
  bool autoScroll = false;

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {



    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: '🎠 Yet_Another_Carousel',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        appBar: AppBar(
          title: Text('🎠 Yet_Another_Carousel'),
          backgroundColor: Colors.blueGrey[900],
        ),
        body: Material(
          color: Colors.blueGrey,
          child: ListView(
            children: <Widget>[
              Carousel(
                children: theseIcons,
                positionCurve: Curves.easeOut,
                scaleCurve: Curves.easeOut,
                fadeOut: true,
                scaleOut: true,
                widthFactor: 2.0,
                scrollTo: wIndex,
              ),
              Divider(height: 2.5,),
              Carousel(
                children: theseIcons,
                positionCurve: Curves.easeOut,
                scaleCurve: Curves.easeOut,
                fadeOut: true,
                widthFactor: 7.5,
                widgetWidth: 0.8,
                widgetHeight: 0.7,
                xPosition: 0.5,
                tailOfLineOffset: Offset(0.0,2.5),
                scrollTo: wIndex,
              ),

            ],
          ),
        ),
        bottomNavigationBar: Container(
          height: 50.0,
          color: Colors.blueGrey[900],
          child: Row(
            mainAxisAlignment: MainAxisAlignment.spaceEvenly,
            children: <Widget>[
              RaisedButton(
                color: Colors.blue,
                child: Text('<<',style: TextStyle(fontSize: 24,color: Colors.white),),
                onPressed: (){
                  setState(() {
                    wIndex -= 1;
                    wIndex = wIndex % theseIcons.length;
                    autoScroll = true;
                  });
                  print(wIndex);
                },
              ),
              Text('$wIndex',style: TextStyle(color: Colors.white,fontSize: 24),),
              RaisedButton(
                color: Colors.blue,
                child: Text('>>',style: TextStyle(fontSize: 24,color: Colors.white),),
                onPressed: (){
                  setState(() {
                    wIndex += 1;
                    wIndex = wIndex % theseIcons.length;
                    autoScroll = true;
                  });
                  print(wIndex);
                },
              ),
            ],
          ),
        ),
      ),
    );
  }
}



double HaW = 150.0;


List<Widget> theseIcons = [
  GestureDetector(
    onTap: (){
      print('Red');
    },
    child:   Container(
      height: HaW,
      width: HaW,
      decoration: BD(color:Colors.red),
      alignment: Alignment.center,
      child: Text('Red0', style: TextStyle(fontSize: 24.0,fontWeight: FontWeight.normal),),
    ),
  ),
  GestureDetector(
    onTap: (){
      print('Orange');
    },
    child: Container(
      height: HaW,
      width: HaW,
      decoration: BD(color:Colors.orange),
      alignment: Alignment.center,
      child: Text('Orange1', style: TextStyle(fontSize: 24.0,fontWeight: FontWeight.normal),),
    ),
  ),
  GestureDetector(
    onTap: (){
      print('Yellow');
    },
    child: Container(
      height: HaW,
      width: HaW,
      decoration: BD(color:Colors.yellow),
      alignment: Alignment.center,
      child: Text('Yellow2', style: TextStyle(fontSize: 24.0,fontWeight: FontWeight.normal),),
    ),
  ),
  GestureDetector(
    onTap: (){
      print('lightGreen');
    },
    child: Container(
      height: HaW,
      width: HaW,
      decoration: BD(color:Colors.lightGreen),
      alignment: Alignment.center,
      child: Text('LightGreen3', style: TextStyle(fontSize: 24.0,fontWeight: FontWeight.normal),),
    ),
  ),
  GestureDetector(
    onTap: (){
      print('Green');
    },
    child: Container(
      height: HaW,
      width: HaW,
      decoration: BD(color:Colors.green),
      alignment: Alignment.center,
      child: Text('Green4', style: TextStyle(fontSize: 24.0,fontWeight: FontWeight.normal),),
    ),
  ),
  GestureDetector(
    onTap: (){
      print('Teal');
    },
    child: Container(
      height: HaW,
      width: HaW,
      decoration: BD(color:Colors.teal),
      alignment: Alignment.center,
      child: Text('Teal5', style: TextStyle(fontSize: 24.0,fontWeight: FontWeight.normal),),
    ),
  ),
  GestureDetector(
    onTap: (){
      print('Blue');
    },
    child: Container(
      height: HaW,
      width: HaW,
      decoration: BD(color:Colors.blue),
      alignment: Alignment.center,
      child: Text(
        'Blue6',
        textAlign: TextAlign.center,
        style: TextStyle(fontSize: 24.0,fontWeight: FontWeight.normal),
      ),
    ),
  ),
  GestureDetector(
    onTap: (){
      print('Indigo');
    },
    child: Container(
      height: HaW,
      width: HaW,
      decoration: BD(color:Colors.indigo),
      alignment: Alignment.center,
      child: Text(
        'Indigo7',
        textAlign: TextAlign.center,
        style: TextStyle(fontSize: 24.0,fontWeight: FontWeight.normal),
      ),
    ),
  ),
  GestureDetector(
    onTap: (){
      print('Purple');
    },
    child: Container(
      height: HaW,
      width: HaW,
      decoration: BD(color:Colors.purple),
      alignment: Alignment.center,
      child: Text(
        'Purple8',
        textAlign: TextAlign.center,
        style: TextStyle(fontSize: 24.0,fontWeight: FontWeight.normal),
      ),
    ),
  ),
  GestureDetector(
    onTap: (){
      print('White');
    },
    child: Container(
      height: HaW,
      width: HaW,
      decoration: BD(color:Colors.white),
      alignment: Alignment.center,
      child: Text(
        'White9',
        textAlign: TextAlign.center,
        style: TextStyle(fontSize: 24.0,fontWeight: FontWeight.normal),
      ),
    ),
  ),
  GestureDetector(
    onTap: (){
      print('Black');
    },
    child: Container(
      height: HaW,
      width: HaW,
      decoration: BD(color:Colors.black),
      alignment: Alignment.center,
      child: Text(
        'Black10',
        textAlign: TextAlign.center,
        style: TextStyle(fontSize: 24.0,color: Colors.white,fontWeight: FontWeight.normal),
      ),
    ),
  ),
  //
//  GestureDetector(
//    onTap: (){
//      print('Red');
//    },
//    child:   Container(
//      height: HaW,
//      width: HaW,
//      decoration: BD(color:Colors.red),
//      alignment: Alignment.center,
//      child: Text('Red0', style: TextStyle(fontSize: 24.0,fontWeight: FontWeight.normal),),
//    ),
//  ),
//  GestureDetector(
//    onTap: (){
//      print('Orange');
//    },
//    child: Container(
//      height: HaW,
//      width: HaW,
//      decoration: BD(color:Colors.orange),
//      alignment: Alignment.center,
//      child: Text('Orange1', style: TextStyle(fontSize: 24.0,fontWeight: FontWeight.normal),),
//    ),
//  ),
//  GestureDetector(
//    onTap: (){
//      print('Yellow');
//    },
//    child: Container(
//      height: HaW,
//      width: HaW,
//      decoration: BD(color:Colors.yellow),
//      alignment: Alignment.center,
//      child: Text('Yellow2', style: TextStyle(fontSize: 24.0,fontWeight: FontWeight.normal),),
//    ),
//  ),
//  GestureDetector(
//    onTap: (){
//      print('lightGreen');
//    },
//    child: Container(
//      height: HaW,
//      width: HaW,
//      decoration: BD(color:Colors.lightGreen),
//      alignment: Alignment.center,
//      child: Text('LightGreen3', style: TextStyle(fontSize: 24.0,fontWeight: FontWeight.normal),),
//    ),
//  ),
//  GestureDetector(
//    onTap: (){
//      print('Green');
//    },
//    child: Container(
//      height: HaW,
//      width: HaW,
//      decoration: BD(color:Colors.green),
//      alignment: Alignment.center,
//      child: Text('Green4', style: TextStyle(fontSize: 24.0,fontWeight: FontWeight.normal),),
//    ),
//  ),
//  GestureDetector(
//    onTap: (){
//      print('Teal');
//    },
//    child: Container(
//      height: HaW,
//      width: HaW,
//      decoration: BD(color:Colors.teal),
//      alignment: Alignment.center,
//      child: Text('Teal5', style: TextStyle(fontSize: 24.0,fontWeight: FontWeight.normal),),
//    ),
//  ),
//  GestureDetector(
//    onTap: (){
//      print('Blue');
//    },
//    child: Container(
//      height: HaW,
//      width: HaW,
//      decoration: BD(color:Colors.blue),
//      alignment: Alignment.center,
//      child: Text(
//        'Blue6',
//        textAlign: TextAlign.center,
//        style: TextStyle(fontSize: 24.0,fontWeight: FontWeight.normal),
//      ),
//    ),
//  ),
//  GestureDetector(
//    onTap: (){
//      print('Indigo');
//    },
//    child: Container(
//      height: HaW,
//      width: HaW,
//      decoration: BD(color:Colors.indigo),
//      alignment: Alignment.center,
//      child: Text(
//        'Indigo7',
//        textAlign: TextAlign.center,
//        style: TextStyle(fontSize: 24.0,fontWeight: FontWeight.normal),
//      ),
//    ),
//  ),
//  GestureDetector(
//    onTap: (){
//      print('Purple');
//    },
//    child: Container(
//      height: HaW,
//      width: HaW,
//      decoration: BD(color:Colors.purple),
//      alignment: Alignment.center,
//      child: Text(
//        'Purple8',
//        textAlign: TextAlign.center,
//        style: TextStyle(fontSize: 24.0,fontWeight: FontWeight.normal),
//      ),
//    ),
//  ),
//  GestureDetector(
//    onTap: (){
//      print('White');
//    },
//    child: Container(
//      height: HaW,
//      width: HaW,
//      decoration: BD(color:Colors.white),
//      alignment: Alignment.center,
//      child: Text(
//        'White9',
//        textAlign: TextAlign.center,
//        style: TextStyle(fontSize: 24.0,fontWeight: FontWeight.normal),
//      ),
//    ),
//  ),
//  GestureDetector(
//    onTap: (){
//      print('Black');
//    },
//    child: Container(
//      height: HaW,
//      width: HaW,
//      decoration: BD(color:Colors.black),
//      alignment: Alignment.center,
//      child: Text(
//        'Black10',
//        textAlign: TextAlign.center,
//        style: TextStyle(fontSize: 24.0,color: Colors.white,fontWeight: FontWeight.normal),
//      ),
//    ),
//  ),
];


BoxDecoration BD({color}){
  return BoxDecoration(
    color: color,
    borderRadius: BorderRadius.circular(5.0),
    boxShadow: [
      new BoxShadow(
        color: Colors.black.withOpacity(0.50),
        offset: new Offset(2.5, 2.5),
        blurRadius: 10.0,
      )
    ],
  );
}

//class Carousel extends StatefulWidget {
//  Carousel({
//    this.widgetList
//});
//
//  List<Widget> widgetList;
//
//  @override
//  _CarouselState createState() => _CarouselState();
//}

//class _CarouselState extends State<Carousel> with SingleTickerProviderStateMixin {
//
//
//  double shift = 0.0;
////  double width = 0.0;
//  Size size = Size(1.0, 1.0);
//  double itemOffset = 4.0;
//  GlobalKey _key = GlobalKey();
//
//
//  List<Widget> rebuildWidgetList()
//  {
//    List<Widget> finalWidgetList = [];
//
//    for (int i = 0; i < widget.widgetList.length ;i++)
////    for (int i = widget.widgetList.length - 1; i >= 0; i-- )
//    {
//      double ratio = i/widget.widgetList.length;
//      ratio *= itemOffset;
//      double negRatio = 1 - ratio;
//
////      print(i);
//
////      ratio = ratio + shift;
//
////      int w = widget.widgetList.length - i -1 ;
////      print(w);
//
////      double RS = (ratio+shift)%1.0;
//      double RS = (ratio+shift).clamp(0.0, 1.0);
//      double p = Curves.decelerate.transform(RS);
//
////      double nRS = (shift)%1.0;
//      double s = 1- Curves.decelerate.flipped.transform(RS);
//      s = 1- Curves.decelerate.flipped.transform(RS);
////      s = s.clamp(0.0,double.infinity) ;
//
//
//      finalWidgetList.add(
//          Transform(
//            transform: Matrix4.rotationY(1-s)..
////              translate(p * size.width, p * size.height * 0.25)
//              translate(
//                  p * (size.width + 500) + (10.0 * ratio) - 500.0,
//                  p * size.height * 0.25)
//              ..scale(s),
//            child: Container(
//              height: size.height * 0.70 * s ,
//              width: size.width * 0.80 * s ,
////            child: FittedBox(
////              fit: BoxFit.fitHeight,
//                child: widget.widgetList[i],
////            ),
//            ),
//          ),
//      );
//
//
////      finalWidgetList.add(
////        Positioned(
//////          left: ratio * 250.0 + 50.0 ,
//////          left: (ratio * (0.5 - (ratio * 0.1))) * MediaQuery.of(context).size.width + 50.0,
//////        left:  p * (size.width + 50) + (100.0 * ratio) + 10.0,
////        left:  p * (size.width + 50 ) + (size.width * 0.10 - (0.50 * widget.widgetList.length)) + (size.width * 0.10 * (1/widget.widgetList.length) * ratio) ,
////          child: Container(
//////            height: size.height * 0.80 * s ,
//////            width: size.width * 0.80 * s ,
////////            child: FittedBox(
////////              fit: BoxFit.fitHeight,
//////              child: widget.widgetList[i],
////////            ),
////          ),
////        )
////      );
//    }
//
//    return finalWidgetList;
//  }
//
//
//  _getSize(){
//    RenderBox RB = _key.currentContext.findRenderObject();
//    size = RB.size;
//
////    print(size);
//  }
//
//  AnimationController _controller;
//  Animation<double> _contentAnimation;
//
//  @override
//  void initState() {
//    // TODO: implement initState
//    WidgetsBinding.instance.addPostFrameCallback(_afterLayout);
//    super.initState();
//
//    shift = (itemOffset/theseIcons.length) * -2;
//    rebuildWidgetList();
//
//    _controller = AnimationController(
//      duration: Duration(milliseconds: 500),
//      vsync: this,
//    );
//
//    _controller.addListener((){
//      setState(() {
////        print(_controller.value);
//
//        print(Curves.easeInOut.transform(1- _controller.value));
//        prevDelta = prevDelta * Curves.ease.transform(1- _controller.value);
//        shift += prevDelta;
//        print('prevDelta $prevDelta');
//
////        _onDragEnd();
////        print(animation.value);
//      });
//    });
//  }
//  _afterLayout(_) {
//    _getSize();
//  }
//
////  double _velocity = 0.0;
////  _onDragEnd(){
////    setState(() {
////      print(_velocity * 0.0005);
////      shift += Curves.ease.flipped.transform(controller.value) * 0.25 * _velocity.sign ;
////      shift = shift.clamp(-1.0 * itemOffset, 1.0);
//////      _velocity = _velocity * 0.90;
////      rebuildWidgetList();
////    });
////  }
//
//  double prevDelta = 0.0;
//
//
//  @override
//  void dispose() {
//    _controller.dispose();
//    super.dispose();
//  }
//
//  @override
//  Widget build(BuildContext context) {
//
//    rebuildWidgetList();
//
//    return GestureDetector(
//      onHorizontalDragUpdate: (value){
//        setState(() {
//
//          prevDelta = value.primaryDelta/size.width;
//          shift += prevDelta;
//          shift = shift.clamp(-1.0 * itemOffset, 1.0);
//
//          print(shift);
//        });
//      },
//      onTap: (){prevDelta=0.0;},
//      onHorizontalDragStart: (value){prevDelta=0.0;},
//      onHorizontalDragEnd: (value){
//        print('prevDelta $prevDelta');
//        _controller.duration = Duration(milliseconds: (prevDelta*100000).abs().round());
//        _controller.forward(from: 0.0);
//      },
//      child: Container(
//        key: _key,
//        height: 250.0,
//        width: double.infinity,
//        color: Colors.grey[900],
//          child: Stack(
//            overflow: Overflow.clip,
//            alignment: AlignmentDirectional.center,
//            children: rebuildWidgetList(),
////        children: theseIcons,
//          ),
//      ),
//    );
//  }
//
//
//}

